File: server/src/main/java/net/lumamc/web/model/SkinRenderer.java
package net.lumamc.web.model;

import java.awt.image.BufferedImage;
import java.util.Arrays;

public final class SkinRenderer {

    private final BlockbenchModel model;

    public SkinRenderer(BlockbenchModel model) {
        this.model = model;
    }

    public BufferedImage render(BufferedImage skin, RenderOptions opt) {
        int ss = Math.max(1, opt.supersample);
        int w = opt.width * ss;
        int h = opt.height * ss;

        float[] zbuf = new float[w * h];
        Arrays.fill(zbuf, Float.NEGATIVE_INFINITY);

        int[] pixels = new int[w * h];
        Arrays.fill(pixels, opt.background.getRGB());

        Camera cam = new Camera(model, opt, w, h);

        int texW = skin.getWidth(), texH = skin.getHeight();
        double uScale = texW / model.uvWidth();
        double vScale = texH / model.uvHeight();

        for (BlockbenchModel.Triangle t : model.triangles()) {
            shadeAndRasterize(t, cam, skin, texW, texH, uScale, vScale, pixels, zbuf, w, h, opt.alphaCutoff);
        }

        BufferedImage hi = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
        hi.setRGB(0, 0, w, h, pixels, 0, w);

        BufferedImage out = (ss == 1) ? hi : downsample(hi, opt.width, opt.height, ss);
        if (opt.trim) {
            out = trim(out, opt.trimPadding);
        }
        return out;
    }

    // Crop away the fully transparent border around the model. Padding adds
    // that many transparent pixels back on each side.
    private static BufferedImage trim(BufferedImage img, int padding) {
        int w = img.getWidth();
        int h = img.getHeight();

        int minX = w, minY = h, maxX = -1, maxY = -1;
        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                int alpha = (img.getRGB(x, y) >>> 24) & 0xFF;
                if (alpha == 0) {
                    continue;
                }
                minX = Math.min(minX, x);
                maxX = Math.max(maxX, x);
                minY = Math.min(minY, y);
                maxY = Math.max(maxY, y);
            }
        }

        if (maxX < minX || maxY < minY) {
            return img;
        }

        int cropW = maxX - minX + 1;
        int cropH = maxY - minY + 1;
        int outW = cropW + padding * 2;
        int outH = cropH + padding * 2;

        BufferedImage out = new BufferedImage(outW, outH, BufferedImage.TYPE_INT_ARGB);
        for (int y = 0; y < cropH; y++) {
            for (int x = 0; x < cropW; x++) {
                out.setRGB(x + padding, y + padding, img.getRGB(minX + x, minY + y));
            }
        }
        return out;
    }

    private void shadeAndRasterize(BlockbenchModel.Triangle t, Camera cam,
                                   BufferedImage skin, int texW, int texH,
                                   double uScale, double vScale,
                                   int[] pixels, float[] zbuf, int w, int h,
                                   int alphaCutoff) {
        double[] pa = cam.project(t.a);
        double[] pb = cam.project(t.b);
        double[] pc = cam.project(t.c);

        double brightness = faceBrightness(t.a, t.b, t.c);

        double x0 = pa[0], y0 = pa[1];
        double x1 = pb[0], y1 = pb[1];
        double x2 = pc[0], y2 = pc[1];

        double den = (y1 - y2) * (x0 - x2) + (x2 - x1) * (y0 - y2);
        if (Math.abs(den) < 1e-9) {
            return;
        }

        int minX = (int) Math.max(0, Math.floor(Math.min(x0, Math.min(x1, x2))));
        int maxX = (int) Math.min(w - 1, Math.ceil(Math.max(x0, Math.max(x1, x2))));
        int minY = (int) Math.max(0, Math.floor(Math.min(y0, Math.min(y1, y2))));
        int maxY = (int) Math.min(h - 1, Math.ceil(Math.max(y0, Math.max(y1, y2))));

        for (int py = minY; py <= maxY; py++) {
            for (int px = minX; px <= maxX; px++) {
                double fx = px + 0.5, fy = py + 0.5;
                double l0 = ((y1 - y2) * (fx - x2) + (x2 - x1) * (fy - y2)) / den;
                double l1 = ((y2 - y0) * (fx - x2) + (x0 - x2) * (fy - y2)) / den;
                double l2 = 1 - l0 - l1;
                if (l0 < -1e-4 || l1 < -1e-4 || l2 < -1e-4) {
                    continue;
                }

                float z = (float) (l0 * pa[2] + l1 * pb[2] + l2 * pc[2]);
                int idx = py * w + px;
                if (z <= zbuf[idx]) {
                    continue;
                }

                double u = l0 * t.uvA[0] + l1 * t.uvB[0] + l2 * t.uvC[0];
                double v = l0 * t.uvA[1] + l1 * t.uvB[1] + l2 * t.uvC[1];
                int tx = clamp((int) Math.floor(u * uScale), 0, texW - 1);
                int ty = clamp((int) Math.floor(v * vScale), 0, texH - 1);

                int argb = skin.getRGB(tx, ty);
                int alpha = (argb >>> 24) & 0xFF;
                if (alpha < alphaCutoff) {
                    continue;
                }

                int r = (int) Math.round(((argb >> 16) & 0xFF) * brightness);
                int g = (int) Math.round(((argb >> 8) & 0xFF) * brightness);
                int b = (int) Math.round((argb & 0xFF) * brightness);

                zbuf[idx] = z;
                pixels[idx] = (0xFF << 24) | (r << 16) | (g << 8) | b;
            }
        }
    }

    // Fixed brightness per face direction, like vanilla Minecraft.
    private static double faceBrightness(double[] a, double[] b, double[] c) {
        double ux = b[0] - a[0], uy = b[1] - a[1], uz = b[2] - a[2];
        double vx = c[0] - a[0], vy = c[1] - a[1], vz = c[2] - a[2];
        double nx = uy * vz - uz * vy;
        double ny = uz * vx - ux * vz;
        double nz = ux * vy - uy * vx;

        double len = Math.sqrt(nx * nx + ny * ny + nz * nz);
        if (len < 1e-9) {
            return 0.8;
        }
        nx /= len;
        ny /= len;
        nz /= len;

        double ax = Math.abs(nx), ay = Math.abs(ny), az = Math.abs(nz);
        if (ay >= ax && ay >= az) {
            return ny > 0 ? 1.00 : 0.50;
        }
        if (az >= ax) {
            return 0.80;
        }
        return 0.60;
    }

    private static int clamp(int v, int lo, int hi) {
        if (v < lo) {
            return lo;
        }
        if (v > hi) {
            return hi;
        }
        return v;
    }

    // Average each block of supersampled pixels back down, keeping alpha right.
    private static BufferedImage downsample(BufferedImage hi, int outW, int outH, int ss) {
        BufferedImage out = new BufferedImage(outW, outH, BufferedImage.TYPE_INT_ARGB);
        for (int y = 0; y < outH; y++) {
            for (int x = 0; x < outW; x++) {
                long r = 0, g = 0, b = 0, a = 0;
                for (int dy = 0; dy < ss; dy++) {
                    for (int dx = 0; dx < ss; dx++) {
                        int argb = hi.getRGB(x * ss + dx, y * ss + dy);
                        int pa = (argb >>> 24) & 0xFF;
                        a += pa;
                        r += ((argb >> 16) & 0xFF) * pa;
                        g += ((argb >> 8) & 0xFF) * pa;
                        b += (argb & 0xFF) * pa;
                    }
                }

                int n = ss * ss;
                int avgA = (int) (a / n);
                int rr, gg, bb;
                if (a == 0) {
                    rr = 0;
                    gg = 0;
                    bb = 0;
                } else {
                    rr = (int) (r / a);
                    gg = (int) (g / a);
                    bb = (int) (b / a);
                }
                out.setRGB(x, y, (avgA << 24) | (rr << 16) | (gg << 8) | bb);
            }
        }
        return out;
    }


    private static final class Camera {
        private final double cYaw, sYaw, cPitch, sPitch;
        private final double cx, cy, cz;
        private final double scale;
        private final double halfW, halfH;

        Camera(BlockbenchModel model, RenderOptions opt, int w, int h) {
            // The face sits on the -Z side, so add 180 to make yaw 0 the front.
            double yaw = opt.yaw + 180;
            this.cYaw = Math.cos(Math.toRadians(yaw));
            this.sYaw = Math.sin(Math.toRadians(yaw));
            this.cPitch = Math.cos(Math.toRadians(opt.pitch));
            this.sPitch = Math.sin(Math.toRadians(opt.pitch));
            this.halfW = w / 2.0;
            this.halfH = h / 2.0;

            double minX = Double.MAX_VALUE, minY = Double.MAX_VALUE, minZ = Double.MAX_VALUE;
            double maxX = -Double.MAX_VALUE, maxY = -Double.MAX_VALUE, maxZ = -Double.MAX_VALUE;
            for (BlockbenchModel.Triangle t : model.triangles()) {
                for (double[] p : new double[][]{t.a, t.b, t.c}) {
                    minX = Math.min(minX, p[0]);
                    maxX = Math.max(maxX, p[0]);
                    minY = Math.min(minY, p[1]);
                    maxY = Math.max(maxY, p[1]);
                    minZ = Math.min(minZ, p[2]);
                    maxZ = Math.max(maxZ, p[2]);
                }
            }

            this.cx = (minX + maxX) / 2;
            this.cy = (minY + maxY) / 2;
            this.cz = (minZ + maxZ) / 2;

            if (opt.scale > 0) {
                this.scale = opt.scale * (w / (double) opt.width);
            } else {
                double ex = 0, ey = 0;
                double[][] corners = {
                    {minX, minY, minZ}, {maxX, minY, minZ}, {minX, maxY, minZ}, {maxX, maxY, minZ},
                    {minX, minY, maxZ}, {maxX, minY, maxZ}, {minX, maxY, maxZ}, {maxX, maxY, maxZ}};
                for (double[] c : corners) {
                    double[] rotated = rotateOnly(c);
                    ex = Math.max(ex, Math.abs(rotated[0]));
                    ey = Math.max(ey, Math.abs(rotated[1]));
                }
                double usableW = w * (1 - 2 * opt.fitMargin);
                double usableH = h * (1 - 2 * opt.fitMargin);
                this.scale = Math.min(usableW / (2 * ex), usableH / (2 * ey));
            }
        }

        private double[] rotateOnly(double[] p) {
            double x = p[0] - cx, y = p[1] - cy, z = p[2] - cz;
            double nx = x * cYaw + z * sYaw;
            double nz = -x * sYaw + z * cYaw;
            double ny = y * cPitch - nz * sPitch;
            nz = y * sPitch + nz * cPitch;
            return new double[]{nx, ny, nz};
        }

        double[] project(double[] p) {
            double[] r = rotateOnly(p);
            return new double[]{halfW + r[0] * scale, halfH - r[1] * scale, r[2]};
        }
    }
}

File: server/src/main/kotlin/net/lumamc/web/model/SkinRenderer.kt
package net.lumamc.web.model

import java.awt.image.BufferedImage
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin
import kotlin.math.sqrt

class SkinRenderer(private val model: BlockbenchModel) {

    fun render(skin: BufferedImage, options: RenderOptions): BufferedImage {
        val supersample = max(1, options.supersample)
        val width = options.width * supersample
        val height = options.height * supersample
        val depthBuffer = FloatArray(width * height) { Float.NEGATIVE_INFINITY }
        val pixels = IntArray(width * height) { options.background.rgb }
        val camera = Camera(model, options, width, height)

        val textureWidth = skin.width
        val textureHeight = skin.height
        val horizontalTextureScale = textureWidth / model.uvWidth()
        val verticalTextureScale = textureHeight / model.uvHeight()

        for (triangle in model.triangles()) {
            shadeAndRasterize(
                triangle,
                camera,
                skin,
                textureWidth,
                textureHeight,
                horizontalTextureScale,
                verticalTextureScale,
                pixels,
                depthBuffer,
                width,
                height,
                options.alphaCutoff
            )
        }

        val highResolutionImage = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)
        highResolutionImage.setRGB(0, 0, width, height, pixels, 0, width)

        var output = if (supersample == 1) {
            highResolutionImage
        } else {
            downsample(highResolutionImage, options.width, options.height, supersample)
        }
        if (options.trim) {
            output = trim(output, options.trimPadding)
        }
        return output
    }

    private fun shadeAndRasterize(
        triangle: BlockbenchModel.Triangle,
        camera: Camera,
        skin: BufferedImage,
        textureWidth: Int,
        textureHeight: Int,
        horizontalTextureScale: Double,
        verticalTextureScale: Double,
        pixels: IntArray,
        depthBuffer: FloatArray,
        width: Int,
        height: Int,
        alphaCutoff: Int
    ) {
        val projectedA = camera.project(triangle.a)
        val projectedB = camera.project(triangle.b)
        val projectedC = camera.project(triangle.c)
        val brightness = faceBrightness(triangle.a, triangle.b, triangle.c)

        val x0 = projectedA[0]
        val y0 = projectedA[1]
        val x1 = projectedB[0]
        val y1 = projectedB[1]
        val x2 = projectedC[0]
        val y2 = projectedC[1]
        val denominator = (y1 - y2) * (x0 - x2) + (x2 - x1) * (y0 - y2)
        if (abs(denominator) < 1e-9) {
            return
        }

        val minimumX = max(0.0, floor(min(x0, min(x1, x2)))).toInt()
        val maximumX = min((width - 1).toDouble(), ceil(max(x0, max(x1, x2)))).toInt()
        val minimumY = max(0.0, floor(min(y0, min(y1, y2)))).toInt()
        val maximumY = min((height - 1).toDouble(), ceil(max(y0, max(y1, y2)))).toInt()

        for (pixelY in minimumY..maximumY) {
            for (pixelX in minimumX..maximumX) {
                val x = pixelX + 0.5
                val y = pixelY + 0.5
                val weight0 = ((y1 - y2) * (x - x2) + (x2 - x1) * (y - y2)) / denominator
                val weight1 = ((y2 - y0) * (x - x2) + (x0 - x2) * (y - y2)) / denominator
                val weight2 = 1 - weight0 - weight1
                if (weight0 < -1e-4 || weight1 < -1e-4 || weight2 < -1e-4) {
                    continue
                }

                val depth = (weight0 * projectedA[2] + weight1 * projectedB[2] + weight2 * projectedC[2]).toFloat()
                val pixelIndex = pixelY * width + pixelX
                if (depth <= depthBuffer[pixelIndex]) {
                    continue
                }

                val u = weight0 * triangle.uvA[0] + weight1 * triangle.uvB[0] + weight2 * triangle.uvC[0]
                val v = weight0 * triangle.uvA[1] + weight1 * triangle.uvB[1] + weight2 * triangle.uvC[1]
                val textureX = clamp(floor(u * horizontalTextureScale).toInt(), 0, textureWidth - 1)
                val textureY = clamp(floor(v * verticalTextureScale).toInt(), 0, textureHeight - 1)
                val color = skin.getRGB(textureX, textureY)
                val alpha = color ushr 24 and 0xFF
                if (alpha < alphaCutoff) {
                    continue
                }

                val red = (((color shr 16) and 0xFF) * brightness).roundToInt()
                val green = (((color shr 8) and 0xFF) * brightness).roundToInt()
                val blue = ((color and 0xFF) * brightness).roundToInt()

                depthBuffer[pixelIndex] = depth
                pixels[pixelIndex] = (0xFF shl 24) or (red shl 16) or (green shl 8) or blue
            }
        }
    }

    private class Camera(model: BlockbenchModel, options: RenderOptions, width: Int, height: Int) {
        private val cosineYaw: Double
        private val sineYaw: Double
        private val cosinePitch: Double
        private val sinePitch: Double
        private val centerX: Double
        private val centerY: Double
        private val centerZ: Double
        private val scale: Double
        private val halfWidth = width / 2.0
        private val halfHeight = height / 2.0

        init {
            val yaw = options.yaw + 180
            cosineYaw = cos(Math.toRadians(yaw))
            sineYaw = sin(Math.toRadians(yaw))
            cosinePitch = cos(Math.toRadians(options.pitch))
            sinePitch = sin(Math.toRadians(options.pitch))

            var minimumX = Double.MAX_VALUE
            var minimumY = Double.MAX_VALUE
            var minimumZ = Double.MAX_VALUE
            var maximumX = -Double.MAX_VALUE
            var maximumY = -Double.MAX_VALUE
            var maximumZ = -Double.MAX_VALUE
            for (triangle in model.triangles()) {
                for (point in arrayOf(triangle.a, triangle.b, triangle.c)) {
                    minimumX = min(minimumX, point[0])
                    maximumX = max(maximumX, point[0])
                    minimumY = min(minimumY, point[1])
                    maximumY = max(maximumY, point[1])
                    minimumZ = min(minimumZ, point[2])
                    maximumZ = max(maximumZ, point[2])
                }
            }

            centerX = (minimumX + maximumX) / 2
            centerY = (minimumY + maximumY) / 2
            centerZ = (minimumZ + maximumZ) / 2

            scale = if (options.scale > 0) {
                options.scale * (width / options.width.toDouble())
            } else {
                var horizontalExtent = 0.0
                var verticalExtent = 0.0
                val corners = arrayOf(
                    doubleArrayOf(minimumX, minimumY, minimumZ),
                    doubleArrayOf(maximumX, minimumY, minimumZ),
                    doubleArrayOf(minimumX, maximumY, minimumZ),
                    doubleArrayOf(maximumX, maximumY, minimumZ),
                    doubleArrayOf(minimumX, minimumY, maximumZ),
                    doubleArrayOf(maximumX, minimumY, maximumZ),
                    doubleArrayOf(minimumX, maximumY, maximumZ),
                    doubleArrayOf(maximumX, maximumY, maximumZ)
                )
                for (corner in corners) {
                    val rotated = rotateOnly(corner)
                    horizontalExtent = max(horizontalExtent, abs(rotated[0]))
                    verticalExtent = max(verticalExtent, abs(rotated[1]))
                }
                val usableWidth = width * (1 - 2 * options.fitMargin)
                val usableHeight = height * (1 - 2 * options.fitMargin)
                min(usableWidth / (2 * horizontalExtent), usableHeight / (2 * verticalExtent))
            }
        }

        private fun rotateOnly(point: DoubleArray): DoubleArray {
            val x = point[0] - centerX
            val y = point[1] - centerY
            val z = point[2] - centerZ
            val rotatedX = x * cosineYaw + z * sineYaw
            var rotatedZ = -x * sineYaw + z * cosineYaw
            val rotatedY = y * cosinePitch - rotatedZ * sinePitch
            rotatedZ = y * sinePitch + rotatedZ * cosinePitch
            return doubleArrayOf(rotatedX, rotatedY, rotatedZ)
        }

        fun project(point: DoubleArray): DoubleArray {
            val rotated = rotateOnly(point)
            return doubleArrayOf(
                halfWidth + rotated[0] * scale,
                halfHeight - rotated[1] * scale,
                rotated[2]
            )
        }
    }

    companion object {
        private fun trim(image: BufferedImage, padding: Int): BufferedImage {
            val width = image.width
            val height = image.height
            var minimumX = width
            var minimumY = height
            var maximumX = -1
            var maximumY = -1

            for (y in 0 until height) {
                for (x in 0 until width) {
                    val alpha = image.getRGB(x, y) ushr 24 and 0xFF
                    if (alpha == 0) {
                        continue
                    }
                    minimumX = min(minimumX, x)
                    maximumX = max(maximumX, x)
                    minimumY = min(minimumY, y)
                    maximumY = max(maximumY, y)
                }
            }

            if (maximumX < minimumX || maximumY < minimumY) {
                return image
            }

            val cropWidth = maximumX - minimumX + 1
            val cropHeight = maximumY - minimumY + 1
            val output = BufferedImage(
                cropWidth + padding * 2,
                cropHeight + padding * 2,
                BufferedImage.TYPE_INT_ARGB
            )
            for (y in 0 until cropHeight) {
                for (x in 0 until cropWidth) {
                    output.setRGB(x + padding, y + padding, image.getRGB(minimumX + x, minimumY + y))
                }
            }
            return output
        }

        private fun faceBrightness(a: DoubleArray, b: DoubleArray, c: DoubleArray): Double {
            val vectorUX = b[0] - a[0]
            val vectorUY = b[1] - a[1]
            val vectorUZ = b[2] - a[2]
            val vectorVX = c[0] - a[0]
            val vectorVY = c[1] - a[1]
            val vectorVZ = c[2] - a[2]
            var normalX = vectorUY * vectorVZ - vectorUZ * vectorVY
            var normalY = vectorUZ * vectorVX - vectorUX * vectorVZ
            var normalZ = vectorUX * vectorVY - vectorUY * vectorVX

            val length = sqrt(normalX * normalX + normalY * normalY + normalZ * normalZ)
            if (length < 1e-9) {
                return 0.8
            }
            normalX /= length
            normalY /= length
            normalZ /= length

            val absoluteX = abs(normalX)
            val absoluteY = abs(normalY)
            val absoluteZ = abs(normalZ)
            if (absoluteY >= absoluteX && absoluteY >= absoluteZ) {
                return if (normalY > 0) 1.0 else 0.5
            }
            return if (absoluteZ >= absoluteX) 0.8 else 0.6
        }

        private fun clamp(value: Int, minimum: Int, maximum: Int): Int = when {
            value < minimum -> minimum
            value > maximum -> maximum
            else -> value
        }

        private fun downsample(
            highResolutionImage: BufferedImage,
            outputWidth: Int,
            outputHeight: Int,
            supersample: Int
        ): BufferedImage {
            val output = BufferedImage(outputWidth, outputHeight, BufferedImage.TYPE_INT_ARGB)
            for (y in 0 until outputHeight) {
                for (x in 0 until outputWidth) {
                    var red = 0L
                    var green = 0L
                    var blue = 0L
                    var alpha = 0L
                    for (offsetY in 0 until supersample) {
                        for (offsetX in 0 until supersample) {
                            val color = highResolutionImage.getRGB(
                                x * supersample + offsetX,
                                y * supersample + offsetY
                            )
                            val pixelAlpha = color ushr 24 and 0xFF
                            alpha += pixelAlpha
                            red += ((color shr 16) and 0xFF) * pixelAlpha
                            green += ((color shr 8) and 0xFF) * pixelAlpha
                            blue += (color and 0xFF) * pixelAlpha
                        }
                    }

                    val averageAlpha = (alpha / (supersample * supersample)).toInt()
                    val averageRed: Int
                    val averageGreen: Int
                    val averageBlue: Int
                    if (alpha == 0L) {
                        averageRed = 0
                        averageGreen = 0
                        averageBlue = 0
                    } else {
                        averageRed = (red / alpha).toInt()
                        averageGreen = (green / alpha).toInt()
                        averageBlue = (blue / alpha).toInt()
                    }
                    output.setRGB(
                        x,
                        y,
                        (averageAlpha shl 24) or (averageRed shl 16) or (averageGreen shl 8) or averageBlue
                    )
                }
            }
            return output
        }
    }
}
